sui_core/src/gateway_state.rs

```diff
@@ -502,104 +271,71 @@ where
-        for object_kind in &transaction.input_objects()? {
-            let object_id = object_kind.object_id();
-            let next_sequence_number = account
-                .highest_known_version(&object_id)
-                .unwrap_or_default();
-            fp_ensure!(
-                object_kind.version() >= next_sequence_number,
-                SuiError::UnexpectedSequenceNumber {
-                    object_id,
-                    expected_sequence: next_sequence_number,
-                    given_sequence: object_kind.version(),
-                }
-                .into()
-            );
-        }
-        // Lock the objects in this transaction
-        account.lock_pending_transaction_objects(&transaction)?;
+        transaction.check_signature()?;
+        let transaction_digest = transaction.digest();
+        let input_objects = transaction.input_objects()?;
+        let objects = self.fetch_objects(&input_objects).await?;
+
+        let all_objects =
+            transaction_input_checker::check_locks(&transaction, input_objects, objects)?;
+        let owned_objects = transaction_input_checker::filter_owned_objects(all_objects);
+        self.store.set_transaction_lock(
+            &owned_objects,
+            transaction_digest,
+            Either::Left(transaction.clone()),
+        )?;
 
         let state = Arc::new(futures::lock::Mutex::new(self));
 
-        // How do we handle errors on authority which lock objects?
-        // Currently VM crash can keep objects locked, but we would like to avoid this.
-        // TODO: https://github.com/MystenLabs/sui/issues/349
-        //       https://github.com/MystenLabs/sui/issues/211
-        //       https://github.com/MystenLabs/sui/issues/346
-        let _guard = scopeguard::guard(state.clone(), |store_ref| {
+        // The scope guard ensures that even if we panic inside transaction execution,
+        // we always unlock the objects from the transaction.
+        // TODO: Handle errors when removing transaction locks?
+        // TODO: Since we removed mut, we shouldn't need Arc and mutex?
+        let _guard = scopeguard::guard(state.clone(), |state_ref| {
             // This should work because the main lock has yielded by the time the guard is at unwinding
-            let mut store = store_ref
+            state_ref
                 .try_lock()
-                .expect("Failed to acquire the store lock post TX execution");
-            let account = store
-                .get_or_create_account(transaction.sender_address())
-                .expect("lost the sender's account");
-            account
-                .unlock_pending_transaction_objects(&transaction)
-                .expect("failed to unlock pending transactions");
+                .expect("Failed to acquire the store lock post TX execution")
+                .store
+                .remove_transaction_lock_only(&owned_objects)
```

Comment:
I am a little unclear about when / why we "unlock" objects that were locked? Are we sure this transaction was invalid, and will never be processed?